zapai/core.go

```diff
@@ -88,11 +143,34 @@ func (c *Core) Write(entry zapcore.Entry, fields []zapcore.Field) error {
 	// reset the traceTelemetry in encoder
 	c.enc.setTraceTelemetry(t)
 
+	t.Properties[appNameStr] = c.appName
+	t.Properties[versionStr] = c.appVersion
+	t.Properties[osStr] = runtime.GOOS
+	t.Properties[hostNameKey], _ = os.Hostname()
 	// set caller
 	if entry.Caller.Defined {
 		t.Properties["caller"] = entry.Caller.String()
 	}
 
+	// Acquire read lock to read metadata
+	c.metaRWMutex.RLock()
+	metadata := c.metadata
+	c.metaRWMutex.RUnlock()
+
+	// Check if metadata is populated
+	if metadata.SubscriptionID != "" {
+		t.Tags.User().SetAccountId(metadata.SubscriptionID)
+		t.Tags.User().SetId(metadata.VMName)
+		t.Properties[locationStr] = metadata.Location
+		t.Properties[resourceGroupStr] = metadata.ResourceGroupName
+		t.Properties[vmIDStr] = metadata.VMID
+		t.Properties[vmNameStr] = metadata.VMName
+		t.Properties[vmSizeStr] = metadata.VMSize
+		t.Properties[osVersionStr] = metadata.OSVersion
+		t.Properties[subscriptionIDStr] = subscriptionIDStr
+		t.Tags.Session().SetId(metadata.VMID)
+	}
```

Comment:
why should this be the responsibility of the core, instead of the responsibility of the _user_ of the core?

Reply:
Ya. Was trying to find a central place for it so each user don't need to have it own implementation for collecting VM meta. 
but if we want to use it in in the components of that module (like CNS, CNI, NPM)., the cyclical dependency is a thing

Reply:
Is it possible to do this as a user of the library only? If not, maybe offer a callback mechanism to allow the user of the core to do this? Just an idea... not fully immersed in this to figure out if it's a good one or not.

Reply:
Something like https://github.com/Azure/azure-container-networking/blob/ce5cc41650653287dca7cf96823b3ea66e9bff12/zapai/mappers.go#L50? 🙂 
Setting all of these fields is already possible (though I will concede that xrefing the mappers to figure out what the magic remapped key is is not the best UX), **even the special ones like `.User().SetId()`**. Those tag groups are just convenience setters to the `ai.user.authUserId` key in the same underlying `Properties` map.